### PR TITLE
feat(gear.exe): append slasher logic to middleware

### PR DESCRIPTION
**Pull request by grishasobol**

### CODE REVIEW

**Comment by StackOverflowExcept1on** [comment]
real diff: https://github.com/gear-tech/gear/compare/dn-rename-router-committed-block-fields...gsobol-symbiotic-slasher

File: ethexe/contracts/test/Middleware.t.sol [hunk 1]
```diff
@@ -5,19 +5,25 @@ import {EnumerableMap} from "@openzeppelin/contracts/utils/structs/EnumerableMap
...
 import {IBaseDelegator} from "symbiotic-core/src/interfaces/delegator/IBaseDelegator.sol";
 import {IOperatorSpecificDelegator} from "symbiotic-core/src/interfaces/delegator/IOperatorSpecificDelegator.sol";
+import {IVetoSlasher} from "symbiotic-core/src/interfaces/slasher/IVetoSlasher.sol";
+import {IBaseSlasher} from "symbiotic-core/src/interfaces/slasher/IBaseSlasher.sol";
 
 import {Middleware} from "../src/Middleware.sol";
 import {WrappedVara} from "../src/WrappedVara.sol";
 import {MapWithTimeData} from "../src/libraries/MapWithTimeData.sol";
 
+bytes32 constant REQUEST_SLASH_EVENT_SIGNATURE =
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:24]: place this inside contract

File: ethexe/contracts/test/Middleware.t.sol [hunk 2]
```diff
@@ -5,19 +5,25 @@ import {EnumerableMap} from "@openzeppelin/contracts/utils/structs/EnumerableMap
 import {Time} from "@openzeppelin/contracts/utils/types/Time.sol";
 import {MessageHashUtils} from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
 import {Upgrades} from "openzeppelin-foundry-upgrades/Upgrades.sol";
-
 import {Test, console} from "forge-std/Test.sol";
+import "forge-std/Vm.sol";
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:9]:
```suggestion
import {Vm} from "forge-std/Vm.sol";
```
iirc `vm.` should be available w/o import in tests

File: ethexe/contracts/src/Middleware.sol [hunk 3]
```diff
@@ -204,12 +305,46 @@ contract Middleware {
...
 
+    function requestSlash(SlashData[] calldata data) external _onlyRole(roleSlashRequester) {
+        for (uint256 i; i < data.length; ++i) {
+            SlashData calldata slash_data = data[i];
+            if (!operators.contains(slash_data.operator)) {
+                revert NotRegistredOperator();
+            }
+
+            for (uint256 j; j < slash_data.vaults.length; ++j) {
+                VaultSlashData calldata vault_data = slash_data.vaults[j];
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/src/Middleware.sol:322]:
```suggestion
                VaultSlashData calldata vaultData = slash_data.vaults[j];
```

File: ethexe/contracts/test/Middleware.t.sol (hunk 2 above)
Comment (grishasobol) [reply ethexe/contracts/test/Middleware.t.sol]: done

File: ethexe/contracts/src/Middleware.sol (hunk 3 above)
Comment (grishasobol) [reply ethexe/contracts/src/Middleware.sol]: done

File: ethexe/contracts/test/Middleware.t.sol [hunk 4]
```diff
@@ -21,13 +24,19 @@ import {MapWithTimeData} from "../src/libraries/MapWithTimeData.sol";
 contract MiddlewareTest is Test {
     using MessageHashUtils for address;
 
+    bytes32 public constant REQUEST_SLASH_EVENT_SIGNATURE =
+        keccak256("RequestSlash(uint256,bytes32,address,uint256,uint48,uint48)");
+
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:29]:
```suggestion
```

File: ethexe/contracts/test/Middleware.t.sol [hunk 5]
```diff
@@ -191,92 +203,321 @@ contract MiddlewareTest is Test {
...
+        } else {
+            middleware.requestSlash(slashes);
+        }
+        Vm.Log[] memory logs = vm.getRecordedLogs();
+
+        uint16 k = 0;
+        for (uint256 i = 0; i < logs.length; i++) {
+            Vm.Log memory log = logs[i];
+            bytes32 eventSignature = log.topics[0];
+            if (eventSignature == REQUEST_SLASH_EVENT_SIGNATURE) {
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:517]:
```suggestion
            if (eventSignature == IVetoSlasher.RequestSlash.selector) {
```

Comment (grishasobol) [reply ethexe/contracts/test/Middleware.t.sol]: done

File: ethexe/contracts/test/Middleware.t.sol [hunk 6]
```diff
@@ -39,33 +45,37 @@ contract MiddlewareTest is Test {
...
-        assertEq(uint256(middleware.OPERATOR_GRACE_PERIOD()), eraDuration * 2);
-        assertEq(uint256(middleware.VAULT_GRACE_PERIOD()), eraDuration * 2);
-        assertEq(uint256(middleware.VAULT_MIN_EPOCH_DURATION()), eraDuration * 2);
-        assertEq(middleware.VAULT_FACTORY(), address(sym.vaultFactory()));
-        assertEq(middleware.DELEGATOR_FACTORY(), address(sym.delegatorFactory()));
-        assertEq(middleware.SLASHER_FACTORY(), address(sym.slasherFactory()));
-        assertEq(middleware.OPERATOR_REGISTRY(), address(sym.operatorRegistry()));
-        assertEq(middleware.COLLATERAL(), address(wrappedVara));
-
         sym.networkRegistry().isEntity(address(middleware));
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:74]:
```suggestion
        assertTrue(sym.networkRegistry().isEntity(address(middleware)));
```

File: ethexe/contracts/test/Middleware.t.sol [hunk 7]
```diff
@@ -191,92 +200,321 @@ contract MiddlewareTest is Test {
...
+
+        // Disable operator1 and check operator1 stake is 0
+        _disableOperator(operator1);
+        // Disable is not immediate, so we need to check for the next block ts
+        uint48 ts = uint48(vm.getBlockTimestamp()) + 1;
+        vm.warp(vm.getBlockTimestamp() + 2);
+        assertEq(middleware.getOperatorStakeAt(operator1, ts), 0);
+
+        // Check that operator1 is not in active operators list
+        (address[] memory active_operators, uint256[] memory stakes) = middleware.getActiveOperatorsStakeAt(ts);
```
Comment (StackOverflowExcept1on) [inline ethexe/contracts/test/Middleware.t.sol:259]:
```suggestion
        (address[] memory activeOperators, uint256[] memory stakes) = middleware.getActiveOperatorsStakeAt(ts);
```

**Review by StackOverflowExcept1on — APPROVED** [review]

File: ethexe/contracts/test/Middleware.t.sol (hunk 7 above)
Comment (grishasobol) [reply ethexe/contracts/test/Middleware.t.sol:259]: done

**Pull request merged**